### PR TITLE
Implement federation of abstract fields (via normalization)

### PR DESCRIPTION
**Pull request by dnerdy**
Fixes #328.

This PR implements federation of abstract fields. It leverages operation
normalization added in Khan/graphql-go-tools#4, as outlined in #328 (in
the comment dated Feb 4, 2022). A previous implementation attempt was
made in #334. This implementation is much simpler, due to the
normalization. The PR includes the original test cases as well as some
additional test cases that cover edge cases not handled by the original
implementation.

Specifically, this PR adds support for the situation where a field with
an interface or union type has concrete types that are owned by another
service (or at least have selected fields owned by another service).

Take, for example, the query:
```
query {
  user {
    username
    pets {  # interface Pet
      name
      ... on Dog {
        dogField
      }
      ... on Cat {
        catField
      }
    }
  }
}
```

In this example "user" and "pets" are both resolved by the "user"
service and "name", "dogField" and "catField" are resolved by the "pet"
service. The issue is that the "pets" field is owned by one service but
the fields of the Pet type are owned by another service.

The major changes are:
- The federation root is added to the upstream opertion the first time an inline fragment or field is visited below the planner parent path.
- The "__typename" variable is now dynamic in representations when the root type is abstract.

Depends on #4.

### CODE REVIEW

**Comment by dnerdy** [comment]
@jensneuse, this is another stacked PR.

File: pkg/engine/plan/plan.go [hunk 1]
```diff
@@ -411,6 +411,79 @@ func (v *Visitor) AllowVisitor(kind astvisitor.VisitorKind, ref int, visitor int
 		return true
 	}
 	for _, config := range v.planners {
+		// In the following query:
```
Comment (dnerdy) [inline pkg/engine/plan/plan.go:414]: @csilvers given your recent experience with planner paths, is this comment any clearer, or do you think it could still benefit from some additional context?

Comment (csilvers) [reply pkg/engine/plan/plan.go:414]: It's definitely clearer, but I'm still not sure it's clear!  But my confusion runs pretty deep, and it's probably best to discuss in person if you want to make any more changes.  (To be clear, there's no need to make changes on my account, if jens thinks this is clear.)

**Comment by dnerdy** [comment]
Closing all the old graphql-go-tools PRs. We ended up using Apollo Gateway to create query plans (which can be done statically), which we then execute via a small, custom Go execution engine.
